### PR TITLE
Ng2 translate

### PR DESCRIPTION
**Pull request by sendilkumarn**
Added [ng2-translate](https://github.com/ocombe/ng2-translate)

@deepu105 need to change the logic which loads the constant file with language constants.

### CODE REVIEW

**Comment by sendilkumarn** [comment]
check dad8b19 for what has to done for translation.

File: generators/client-2/templates/src/main/webapp/app/home/home.html [hunk 1]
```diff
@@ -4,7 +4,7 @@
             <span class="hipster img-responsive img-rounded"></span>
         </div>
         <div class="col-md-8">
-            <h1 translate="home.title">Welcome, Java Hipster!</h1>
+            <h1 [jhi-translate]="'home.title' | translate">Welcome, Java Hipster!</h1>
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: why do we need both directive and pipe? cant we do with one?

File: generators/client-2/templates/src/main/webapp/app/layouts/navbar/_navbar.component.ts [hunk 2]
```diff
@@ -23,6 +27,7 @@ export class NavbarComponent implements OnInit {
         private loginService : LoginService,
         <%_ if (enableTranslation){ _%>
         private languageService: <%=jhiPrefixCapitalized%>LanguageService,
+        private translateService: TranslateService,
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/layouts/navbar/_navbar.component.ts]: is it used?

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts [hunk 3]
```diff
@@ -24,12 +34,19 @@ import {
 @NgModule({
     imports: [
         <%=angular2AppName%>SharedLibsModule,
-        <%=angular2AppName%>SharedCommonModule,
+        <%=angular2AppName%>SharedCommonModule<%_ if (enableTranslation){ _%>,
+        TranslateModule.forRoot({
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: can we move this to a provider file and call it here? similar to alert provider

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts [hunk 4]
```diff
@@ -43,15 +60,28 @@ import {
...
+        TranslateModule,
+        <%=jhiPrefixCapitalized%>Translate
     ],
     schemas: [CUSTOM_ELEMENTS_SCHEMA]
 
 })
-export class <%=angular2AppName%>SharedModule {}
+export class <%=angular2AppName%>SharedModule {
+    <%_ if (enableTranslation){ _%>
+     static forRoot(): ModuleWithProviders {
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: why is this required?

File: generators/client-2/templates/src/main/webapp/app/home/home.html (hunk 1 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: cant we use the pipe inside the directive? and also do away with the inner single quotes?

File: generators/client-2/templates/src/main/webapp/app/shared/language/_jhiMissing.translation.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import {MissingTranslationHandler} from 'ng2-translate/ng2-translate';
+
+export class <%=jhiPrefixCapitalized%>MissingTranslationHandler implements MissingTranslationHandler {
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/language/_jhiMissing.translation.ts]: can we rename the file to `jhi-missing-translation-handler.config.ts` ?

File: generators/client-2/templates/src/main/webapp/app/shared/language/_language.service.ts [hunk 6]
```diff
@@ -1,25 +1,37 @@
-import { LANGUAGES } from './language.constants';
 import { Injectable, Inject } from '@angular/core';
 
+import { TranslateService } from 'ng2-translate/ng2-translate';
+
+import { LANGUAGES } from './language.constants';
+import { <%=angular2AppName%>SharedCommonModule } from '../../shared';
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/language/_language.service.ts]: why is the module imported?

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 3 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: I think this should be in the main AppModule as that is the root

Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: Ok I just read the docs, so it can be in a shared module. In that case better add it to the `<%=angular2AppName%>SharedLibsModule` and add the custom translation related directives/pipes in the `SharedCommonModule` as this module imports both those

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts [hunk 7]
```diff
@@ -24,12 +34,19 @@ import {
 @NgModule({
     imports: [
         <%=angular2AppName%>SharedLibsModule,
-        <%=angular2AppName%>SharedCommonModule,
+        <%=angular2AppName%>SharedCommonModule<%_ if (enableTranslation){ _%>,
+        TranslateModule.forRoot({
+          provide: TranslateLoader,
+          useFactory: (http: Http) => new TranslateStaticLoader(http, '/i18n', '.json'),
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: as per docs isnt this the default `TranslateStaticLoader` then do we have to overwrite it?

File: generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  constructor(private el: ElementRef, private renderer: Renderer) { }
+
+  @Input()
+  set jhiTranslate(value: string | TemplateRef<any>) {
+      this._jhiTranslate = value;
+      this.setTranslatedValue();
+  }
+
+  setTranslatedValue() {
+    if (this._jhiTranslate ) {
```
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts]: use service to translate key so that we can avoid the pipe and do `jhi-translate="home.title"` https://github.com/ocombe/ng2-translate#4-use-the-service-or-the-pipe

**Comment by deepu105** [comment]
Good work @sendilkumarn please take a look at my comments

File: generators/client-2/templates/src/main/webapp/app/home/home.html (hunk 1 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: `pipe` is following ng2-translate.
`directive` manually done.

They have a method inside But I don't want to use that and re-engineer the thing.

The api needs `single-quote`.

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 4 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: To make `TranslateService` as a singleton.

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 7 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: no default is `assets/i18n`

File: generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts (hunk 8 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts]: I somehow feel using this will slow things down. I will check the internals

File: generators/client-2/templates/src/main/webapp/app/home/home.html (hunk 1 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: This one is tricky. we have to trigger this directive when the language changes, which will be an overhead.

File: generators/client-2/templates/src/main/webapp/app/layouts/navbar/_navbar.component.ts (hunk 2 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/layouts/navbar/_navbar.component.ts]: removed 👍

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 3 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: This one has some wiring problems. Can you check this part. I don't know, moving this chunk over that file is causing problem.

File: generators/client-2/templates/src/main/webapp/app/shared/language/_jhiMissing.translation.ts (hunk 5 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/language/_jhiMissing.translation.ts]: Done 👍

File: generators/client-2/templates/src/main/webapp/app/shared/language/_language.service.ts (hunk 6 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/language/_language.service.ts]: Removed 👍

File: generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts (hunk 8 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts]: Refer my first comment.

File: generators/client-2/templates/src/main/webapp/app/home/_home.component.ts [hunk 9]
```diff
@@ -14,7 +18,12 @@ export class HomeComponent implements OnInit {
 
     constructor(private principal: Principal,
                 private $state: StateService,
-                private loginService : LoginService) {
+                private loginService : LoginService<%_ if (enableTranslation){ _%>,
+                translateService: TranslateService,
+                <%=jhiPrefix%>LangService: <%=jhiPrefixCapitalized%>LanguageService<%_ } _%>) {
+        <%_ if (enableTranslation){ _%>
+        <%=jhiPrefix%>LangService.setLocation('home');
```
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/home/_home.component.ts]: Done 👍 👏

File: generators/client-2/templates/src/main/webapp/app/home/home.html (hunk 1 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: Ya that would be there with or without the pipe rite?
I would prefer a simpler directive which does everything. we can use the service inside the directive. Also I think its possible to get the directive evaluated during digest cycles

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 3 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: ok

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 4 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: ok

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 7 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: doc says below

> By default, only the TranslateStaticLoader is available. It will search for files in i18n/*.json, if you want you can customize this behavior by changing the default prefix/suffix:

can you check

File: generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts (hunk 8 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/shared/directive/jhi-translate.ts]: I dont think there will be much difference here. The pipe would be using the service most probably so we can just use it directly

File: generators/client-2/index.js [hunk 10]
```diff
@@ -501,6 +501,7 @@ module.exports = JhipsterClientGenerator.extend({
             this.template(ANGULAR_DIR + 'shared/directive/maxbytes.directive.ts', ANGULAR_DIR + 'shared/directive/maxbytes.directive.ts', this, {});
             this.template(ANGULAR_DIR + 'shared/directive/minbytes.directive.ts', ANGULAR_DIR + 'shared/directive/minbytes.directive.ts', this, {});
             this.template(ANGULAR_DIR + 'shared/directive/number-of-bytes.ts', ANGULAR_DIR + 'shared/directive/number-of-bytes.ts', this, {});
+            this.template(ANGULAR_DIR + 'shared/directive/jhi-translate.ts', ANGULAR_DIR + 'shared/directive/jhi-translate.ts', this, {});
```
Comment (deepu105) [inline generators/client-2/index.js]: move this into the language folder

Comment (sendilkumarn) [inline generators/client-2/index.js]: No this has to be in directive moving this will make us to create `language` folder even when there is no translation

File: generators/client-2/templates/src/main/webapp/app/home/home.html (hunk 1 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: I dont like the markup its redundant. Let me see if there is a way to make it work kust with directive

File: generators/client-2/index.js (hunk 10 above)
Comment (sendilkumarn) [inline generators/client-2/index.js]: to return empty so that it will not replace anything. else everywhere in html we have to add in a check

Comment (deepu105) [inline generators/client-2/index.js]: Ok currently we do a regex strip of translate directives and pipes so that if translation is disabled the generated project is clean. We need to do the same here as well. Once migration is complete we can modify the regex to strip these as well. So no need to have the directive added when there is no translation

**Comment by deepu105** [comment]
@sendilkumarn i'm working on this

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 4 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: Yeah Okay.

File: generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts (hunk 3 above)
Comment (sendilkumarn) [inline generators/client-2/templates/src/main/webapp/app/shared/_shared.ng2module.ts]: Okay 👍

File: generators/client-2/templates/src/main/webapp/app/home/home.html (hunk 1 above)
Comment (deepu105) [inline generators/client-2/templates/src/main/webapp/app/home/home.html]: @sendilkumarn the solution for this is too simple and stupid, we just need to think from a stupid perspective 😉 
```js
import { Component } from '@angular/core';

@Component({
  selector: '[jhi-translate],[jhiTranslate]',
  template: '{{ key | translate }}',
  inputs: ['key:jhi-translate']
})
export class JhiTranslate {

  private key: string;
  //TODO add support to pass translate-values and for translate-compile

}
```
